arraycontext/parameter_study/transform.py

```diff
@@ -0,0 +1,602 @@
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from dataclasses import dataclass
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    FrozenSet,
+    Mapping,
+    Sequence,
+    Set,
+    Tuple,
+    Union,
+)
+
+import numpy as np
+import pymbolic.primitives as prim
+from immutabledict import immutabledict
+
+import loopy as lp
+import pytato as pt
+from pytato.array import (
+    Array,
+    AxesT,
+    Axis,
+    AxisPermutation,
+    Concatenate,
+    Einsum,
+    EinsumElementwiseAxis,
+    IndexBase,
+    IndexLambda,
+    Placeholder,
+    Reshape,
+    Roll,
+    ShapeType,
+    Stack,
+)
+from pytato.scalar_expr import IdentityMapper
+from pytato.transform import CopyMapper
+from pytools.tag import Tag, UniqueTag
+
+from arraycontext import PytatoPyOpenCLArrayContext
+from arraycontext.container import is_array_container_type
+from arraycontext.container.traversal import rec_keyed_map_array_container
+from arraycontext.impl.pytato.compile import (
+    LazilyPyOpenCLCompilingFunctionCaller,
+    LeafArrayDescriptor,
+    _ary_container_key_stringifier,
+    _get_arg_id_to_arg_and_arg_id_to_descr,
+    _to_input_for_compiled,
+)
+
+
+ArraysT = Tuple[Array, ...]
+
+
+@dataclass(frozen=True)
+class ParameterStudyAxisTag(UniqueTag):
+    """
+        A tag for acting on axes of arrays.
+        To enable multiple parameter studies on the same variable name
+        specify a different axis number and potentially a different size.
+
+        Currently does not allow multiple variables of different names to be in
+        the same parameter study.
+    """
+    # user_param_study_tag: Tag
+    axis_num: int
+    axis_size: int
+
+
+StudiesT = Tuple[ParameterStudyAxisTag, ...]
+
+
+class ExpansionMapper(CopyMapper):
+
+    def __init__(self, actual_input_shapes: Mapping[str, ShapeType],
+                 actual_input_axes: Mapping[str, FrozenSet[Axis]]):
+        super().__init__()
+        self.actual_input_shapes = actual_input_shapes
+        self.actual_input_axes = actual_input_axes
+
+    def single_predecessor_updates(self, curr_expr: Array,
+                                          new_expr: Array) -> Tuple[ShapeType,
```

Comment:
One of the parameter names should identify that it's talking about the predecessor.